ISSUE #978: decouple metaformat cmd

reddycharan opened the pull request:
Descriptions of the changes in this PR:

- introduce new commands - initnewcluster, nukeexistingcluster, whatisinstanceid
- add testcases initnewcluster and nukeexistingcluster at BookKeeperAdmin level

Master Issue: #978

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java:
```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
+        List<ACL> zkAcls = ZkUtils.getACLs(conf);
+        // Create ledgers root node if not exists
+        if (!ledgerRootExists) {
+            zk.create(conf.getZkLedgersRootPath(), "".getBytes(Charsets.UTF_8), zkAcls, CreateMode.PERSISTENT);
```
Why not just call create() and ignore the EEXIST exception?  This variant has almost the same effect, but suffers from a race if the ledgerRoot is created between the check and here.  Same with available bookies node and the READONLY node.

reddycharan replied:
I didnt touch the implementation logic of this method, I just removed method arg for this method and try block.

reddycharan replied:
for this commit, I would leave as it is

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/FlatLedgerManagerFactory.java:
```diff
@@ -88,13 +88,14 @@ public LedgerUnderreplicationManager newLedgerUnderreplicationManager()
-        for (String child : children) {
-            if (FlatLedgerManager.isSpecialZnode(child)) {
-                continue;
+        try (FlatLedgerManager ledgerManager = (FlatLedgerManager) newLedgerManager()) {
```
Why not make isLedgerParentNode static?

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/HierarchicalLedgerManager.java:
```diff
@@ -86,6 +86,12 @@ protected long getLedgerId(String ledgerPath) throws IOException {
     }
 
+    @Override
+    public boolean isSpecialZnode(String znode) {
```
This would probably have to be legacyLM.isSpecialZnode(znode) || longLM.isSpecialZnode(znode) || super.isSpecialZnode(znode);

reddycharan replied:
as we discussed, I'm merging all of these isSpecialZnode methods into one method, for the sake of simplicity and clarity.

jiazhai commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java:
```diff
@@ -230,4 +241,200 @@ public void testDecommissionForLedgersWithMultipleSegmentsAndNotWriteClosed() th
         bkAdmin.close();
     }
+
+    @Test(timeout = 60000)
```
Since we have a default timeout setting for all tests, it is suggested to remove `timeout` for individual test case.

reddycharan replied:
what is the default time? and what if we want more time?

sijie replied:
- default time is more than 60 seconds
- typically if a test is required more than 60 seconds, then we really need to think about how to improve the test. A lot of frakiness in current bookkeeper tests come from unnecessarily `Thread.sleep` or wait for something to trigger at a given time. we should try to avoid that pattern.

reddycharan replied:
removed timeouts

jiazhai commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LegacyHierarchicalLedgerManagerFactory.java:
```diff
@@ -79,7 +79,7 @@ public LedgerIdGenerator newLedgerIdGenerator() {
 
     @Override
     public LedgerManager newLedgerManager() {
-        return new HierarchicalLedgerManager(conf, zk);
```
oh, the original code seems weird, why it return HierarchicalLedgerManager?

reddycharan replied:
yes, I do see few other issues with 057af8dbce6c08794eb8b46ca52ca13f222d9bbb . Tried to address in this commit and we noticed other issues as well - Issue #1014

jiazhai commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookKeeperAdminTest.java:
```diff
@@ -230,4 +241,200 @@ public void testDecommissionForLedgersWithMultipleSegmentsAndNotWriteClosed() th
+        String readInstanceId = new String(data, UTF_8);
+
+        /*
+         * registrate a RO bookie
```
typo? register

reddycharan replied:
fixed it

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java:
```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
+    }
+
+    @Override
+    public boolean initNewCluster() throws Exception {
```
the logic here is problematic than the regular format. 

in regular format, it is crashing at the middle of a formatting op, you can re-run format to get it eventually done. in your implementation here, you can't. If I am crashing in the middle of a `init` op, I can't simply re-run `init`.

You might want to document this behavior on those two newly introduced command, to make sure people how to run the commands correctly.

sijie replied:
an alternative way to use `zk.txn` to make sure all the path creations are done in a multi op. but this probably is out of this scope, can be done in a separate task.

reddycharan replied:
I added enough logging when these operations fail, for the admin users to know why exactly it failed. Anyhow I'll add little more information to the issue description, regarding usage of these commands.

sijie replied:
yes. we need enough instructions for people to run the command when they fail at the middle of running this command.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java:
```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
+            return true;
+        }
+
+        String availableBookiesPath = conf.getZkAvailableBookiesPath();
```
The logic here is way too complicated. You don't need to read the list the znodes and check if they are still exists. You can just issue a zk txn (don't remember that exactly method, but the following code is trying to demo how it works), if there are still bookies alive, the txn will fail.

```
Multi multi = zk.multi();
multi.delete(zkLedgersRootPath + "/available/readonly")
multi.delete(zkLedgersRootPath + "/available");

multi.execute();
```

sijie replied:
if you do as what I suggested, you don't need to change BookieWatcher.

reddycharan replied:
Here before proceeding with nuking cluster data, I would like to make sure that there are no connected bookies to the cluster and print informative log message for each step/issue. Also, I moved deletions of the ZK nodes to  LedgerManagerFactory class, for the sake better abstraction. Anyhow the changes I made to BookieWatcher are minimal. For now, I feel comfortable to make independent zk operations, because check/delete operations are made at multiple classes with proper validation.

sijie replied:
> Here before proceeding with nuking cluster data, I would like to make sure that there are no connected bookies to the cluster and print informative log message for each step/issue.

the operation that I proposed is doing exactly same purpose as you hope, with much simpler approach and less error prone.

if you really feel the need to read and delete, you should not use BookieWatch, you should just use registration client api, rather than api in bookie watch. because bookie watch is the user of metadata interfaces, the metadata interface implementation should not use those api, this is a cycle dependency. 

These are the apis you should use: 

https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/RegistrationClient.java#L75
https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/RegistrationClient.java#L82

reddycharan replied:
reverting BookieWatcher changes and going to use RegistrationClient API

athanatos commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java:
```diff
@@ -59,10 +70,14 @@
 
+
+
+
```
Stray whitespace?

reddycharan replied:
fixed it

sijie commented on bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/LedgerManagerIteratorTest.java:
```diff
@@ -517,4 +557,20 @@ public void testLedgerManagerFormat() throws Throwable {
     }
+
+    @Test
+    public void hierarchicalLedgerManagerAsyncProcessLedgersTest() throws Throwable {
```
nit: better to follow the naming as other methods here, so it is consistent.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractHierarchicalLedgerManager.java:
```diff
@@ -63,7 +63,11 @@ void asyncProcessLevelNodes(
             @Override
             public void processResult(int rc, String path, Object ctx) {
-                if (rc != Code.OK.intValue()) {
+                if (rc == Code.NONODE.intValue()) {
```
These commits here seem not related to "decouple metaformat cmd". I would suggest they should go to a separate issue.

reddycharan replied:
ok..would move the 3rd commit to a new PR. But Bug1 in Issue #1014, would be fixed in this PR, since it is getting fixed as part of my changes for this PR. But Bug2 in Issue #1014 and testcases for both the bugs will be in new PR (which is currently 3rd commit).

reddycharan commented:
@sijie 

- I removed commit related to Issue #1014 from this PR
- reverted changes made to BookieWatcher, used RegistrationClient instead for getting RW and RO bookies in ZKRegistrationManager
- Will add new commands description/usage to the issue

reddycharan commented:
@sijie you want me to provide more details to the issue description - https://github.com/apache/bookkeeper/issues/978 ?

sijie commented:
@reddycharan no. I mean adding more descriptions in the commands. For example, if I failed when running initNewCluster, what I should do next? For users who don’t write the code, they will never not how to do. Because it is unlike meta format , you can run many times. You need to make instructions to ask people to nuke the cluster and reinit the cluster. Either you add the message at help message, or print this when failure happened. But we need something in the tool to tell users.

reddycharan commented:
done adding suggested description and comments to the command. I think it should be good to go now.

sijie commented:
@reddycharan can you rebase your branch? it should be ready to go.

reddycharan commented:
rebased and checks are looking good

reddycharan commented:
@sijie can you push this PR. Since this change touches quite a few files, I might run into rebasing issue again if it is delayed. Also, I've a bug fix which goes on top of this PR and I'm waiting for this PR to be closed.

I think other PRs of mine are also good to go.

sijie commented:
@reddycharan sure. I would do it now. sorry I was busy yesterday, so didn't get any chances to go through those PRs.
